tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+    event(lock, ProxyEvent::access_attach);
+  }
+
+  void event_access_release() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::access_release);
+  }
+
+  /**
+   * Predicate whether the state machine is in the `access` state.
+   *
+   * This predicate is used to determine whether to grant access to an
+   * underlying object.
+   */
+  [[nodiscard]] bool is_attached() const {
+    return state_ == ProxyState::access || state_ == ProxyState::last_access;
+  }
+
+ public:
+  /**
+   * Order to construct implemented here for exposure through `Proxy`.
+   *
+   * If validation or construction throws, the proxy will enter its error state.
+   * If validation does not succeed, construction will not happen and the proxy
+   * will stay in the `nascent` state.
+   */
+  void construct() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::construct);
+  }
+
+  /**
+   * Order to destroy implemented here for exposure through `Proxy`.
+   */
+  void destroy() {
+    lock_type lock{m_state_, std::defer_lock};
+    event(lock, ProxyEvent::destroy);
+  }
+
+  /**
+   * Order to shut down implemented here for exposure through `Proxy`.
+   */
+  void shutdown() {
+    if constexpr (has_shutdown) {
+      lock_type lock{m_state_, std::defer_lock};
+      event(lock, ProxyEvent::shutdown);
+    } else {
+      destroy();
+    }
+  }
+};
+
+/*
+ * Life cycle concept
+ *
+ * The `LC` template argument of `Proxy` has a number of requirements, some of
+ * which are expressible with C++20 `concept`, some not.
+ *
+ * * The class has a default constructor. This is necessary to allow `Proxy` to
+ *   have a default constructor, since `Proxy` contains a member variable of
+ *   this type.
+ * * The class has a declaration `arguments_type` that holds arguments for the
+ *   constructor during the nascent state of the proxy. The `Proxy` class is
+ *   outside _how_ this class is used. Typically its member variables act like a
+ *   C-style `struct` and are not proscribed by class invariants. There's no
+ *   requirement that the arguments class be this simple. An arguments class can
+ *   do more, but that's outside the purview of `Proxy`.
+ * * The class `LC::arguments_type` has a default constructor.
+ * * The class has a `B validate()` member function that says whether the
+ *   gathered arguments are well-formed, where `B` is a bool-convertible return
+ *   value. This can be anything from a constant return to a full
+ *   pre-construction validation.
+ * * The underlying class has construtor `T(LC::arguments_type&)`.
+ * * The class has a `constexpr bool has_shutdown` member variable that
+ *   indicates whether the class has a `shutdown()` method defined. A class can
+ *   shutdown if it can refuse new operations that would change any internal
+ *   state. This facility exists to allow classes to report error conditions
+ *   gracefully instead of simply disappearing.
```

Comment:
Would it be useful to use template meta programming for any of this?  E.g., to determine if `shutdown()` exists?  (Not sure if much would be gained since LC are bespoke in some sense anyway.)  I would suggest `has_shutdown` be static in LC implementations (though that is not an enforceable requirement from the client).